src/libexpr/tests/coro-gc.cc

```diff
@@ -0,0 +1,147 @@
+#include <gtest/gtest.h>
+#if HAVE_BOEHMGC
+#include <gc/gc.h>
+
+#include "eval.hh"
+#include "serialise.hh"
+
+#endif
+
+
+namespace nix {
+#if HAVE_BOEHMGC
+
+    static void finalizer(void *obj, void *data) {
+        *((bool*)data) = true;
+    }
+
+    static bool* make_witness(volatile void* obj) {
+        /* We can't store the witnesses on the stack,
+           since they might be collected long afterwards */
+        bool* res = (bool*)GC_MALLOC_UNCOLLECTABLE(1);
+        *res = false;
+        GC_register_finalizer((void*)obj, finalizer, res, nullptr, nullptr);
+        return res;
+    }
+
+    // Generate 2 objects, discard one, run gc,
+    // see if one got collected and the other didn't
+    // GC is disabled inside coroutines on __APPLE__
```

Comment:
I think this kind of test is inherently unreliable, particularly the part where it checks that an object got collected. Since Boehm is a conservative collector, you can never be certain that no pointer to an object survives. For instance, you can't really count on an assignment like `do_collect = nullptr;` not being optimized away.

Reply:
You could increase the odds by allocating a large number of individual objects.
If none of them gets collected, we probably have a significant GC issue anyway.